### PR TITLE
[mlir][transform] Fix handling of transitive include in interpreter.

### PR DESCRIPTION
**Pull request by ingomueller-net**
This is a new attempt at fixing transitive includes in the transform dialect interpreter (next to #67241) and a preparation for being able to load multiple transform library files (#67120).

Until now, the interpreter would only load those symbols from the provided library files that were declared in the main transform module. However, sequences in the library may include other sequences on their own. Until now, if such sequences were not *also* declared in the main transform module, the interpreter would fail to resolve them. Forward declaring all of them is undesirable as it defeats the purpose of encapsulation into library modules.

This PR implements a kind of linker for transform scripts to solve this problem. The linker merges all symbols of the library module into the main module before interpreting the latter. Symbols whose names collide are handled as follows: (1) if they are both functions (in the sense of `FunctionOpInterface`) with compatible signatures, one is external, and the other one is public, then they are merged; (2) of one of them is private, that one is renamed; and (3) an error is raised otherwise.

TODO:
- [x] Enable running interpreter several times (see discussion below).
- [x] Fix several smaller issues marked with `XXX`.
- [x] Fix reproducer: since injection of the library is now not idempotent anymore, we should not print the filename argument in the repro call.
- [x] Revise CLI and function parameter documentation and description.

### CODE REVIEW

**Comment by ingomueller-net** [comment]
@ftynse: This is a new attempt at #67241 along the lines of your suggestions. Please take a look. It is worth mentioning that (i) I merged the stages (1) and (2) from your suggestion, and (ii) did not implement the optimization of only including needed symbols.

I have identified one more problem that may require a major change: it is currently not always possible to run the interpreter pass several times. The problem is the following: if the script contains a declaration `@foo` and the library contains a matching definition, then than definition is merge into the transform module upon its first invocation. In the second invocation, however, `@foo` is now a *definition* in the transform module, so merging in the library module leads to a `doubly defined symbol` error. This did not happen before because only the (directly) missing symbols were merged, and `@foo` is not missing after the first invocation anymore.

I see two ways out: (A) only merge in missing symbols (plus their dependencies) as previously and make all imported symbols private, or (B) separate the merging from the interpreter into a separate pass. I have the feeling that Approach (A) would make the merging again more special to the interpreter and possibly somewhat brittle and Approach (B) would be more generic and possibly more performant. (If the interpreter pass is really used several times, linking has to be done each time, while with a dedicated linking pass, it would only be done once.)

I imagine the linking pass from Approach (B) to be called something like `load-transform-dialect-interpreter-module` or similar, take the `transform-library-file-name` and `transform-library-file-name` arguments (as well as their extensions planned in #67120), and then only do the loading and merging/linking (but not interpret the result). The main interpreter pass would then simply run the interpreter on the input module and not bother with any loading or merging. Also, I think I'd implement that separation *first* and then rebase this change on top of it.

What do you think?

**Comment by ftynse** [comment]
Approach B makes sense to me. Generally, anything that keeps the interpreter simpler and more focused is likely a good idea. I suppose it should be sufficient to just move the symbols from other files,normally, it is allowed to have multiple declarations as long as there is a single definition + handle name conflicts between _private_ symbols that may have repeated names in different modules. Then we can ran symbol-dce separately to remove unused symbols.

**Comment by ftynse** [comment]
Implementation wise, we can consider having a "mergeModules" function that takes a list of modules and moves everything into the first one. It can then be exercised from the pass that loads transform library files, but is also reusable.

**Comment by ingomueller-net** [comment]
> Implementation wise, we can consider having a "mergeModules" function that takes a list of modules and moves everything into the first one. It can then be exercised from the pass that loads transform library files, but is also reusable.

Yeah, good point! Being able to merge several modules is on my TODO list. That would also be required if we wanted to write a generic merger that could replace most of [SPIR-V's `ModuleCombiner`](https://github.com/llvm/llvm-project/blob/d5c8b23b/mlir/include/mlir/Dialect/SPIRV/Linking/ModuleCombiner.h#L70). But I'll leaver that to that after the remaining pending changes...

**Comment by ingomueller-net** [comment]
Yes, but you can do (slightly) better by being careful when to (re)construct which symbol table(s)...

**Comment by ingomueller-net** [comment]
I think I changed my mind about the question of whether or not loading should be part of the interpreter. At the very least, I see that it is absolutely consistent to think of the current pass as "inject dependencies and interpret result." In this case, it is clear that running that pass twice in a way that tries to inject the dependencies twice does not work. The PR thus currently removes that test. If we decide that injection should be separated from the interpreter or that injection should be done only if necessary (such that it is idempotent), I believe we can do that in some future PR.

Other than that, I have overhauled the initial commit and think that it is ready for review now.

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:330]: I'm not very happy about this low-level renaming detail here. I suppose if this moves to `SymbolTable.h/cpp` as a utility that shares the renaming logic with the table, it will be okay. Only this part, not the the use replacement.

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:334]: Nit: looks like the debug line starts with a comma for no reason.

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:316]: Nit: can't we just use the symbol table to resolve the name and obtain `otherOp` instead of passing it in?

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:360]: This condition is unclear to me. Why can we merge an external function into another external function? Neither has a body... Why should the other function be public? Is this about merging _declarations_?

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:375]: Nit: I wonder if we can construct this in the dialect initialization and have the dialect give us the attributes directly.

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:423]: Why do we need to pass an owning reference here?

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:439]: Nit: use `auto &&` for zippers.

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:432]: I don't see the point of sharing the counter across different symbol names. Starting at 0 on each collision would produce more readable names.

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:490]: Nit: prefer the `emitError(loc) << foo << bar` format. Twines are always tricky.

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:441]: This place is the rare place to explicitly specify the number of stack elements in the vector.

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:500]: I don't think we should be running local verification here. Has it caught any errors?
Maybe we should consider having "merge" as a separate pass that runs before the interpreter so there is an opportunity for the verifier to run in between (and to be disabled for compiler performance-sensitive cases).

**Review by ingomueller-net — COMMENTED** [review]
Thanks a lot, @ftynse, for the detailed feedback! I just pushed a bunch of commits that should address your comments (modulo the questions I am submitting here).

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [reply mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp]: LLVM doesn't like ravioli commits.

File: mlir/include/mlir/Dialect/Transform/IR/TransformDialect.td
Comment (ftynse) [inline mlir/include/mlir/Dialect/Transform/IR/TransformDialect.td:45]: Nit: this comment is rendered, e.g., for autocompletion, where there is no "above" context. Prefer something more explanatory.

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:53]: MLIR has a preference for using static functions.

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:301]: Nit: declare as static.

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [reply mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:302]: A more fitting name could be beneficial, but I can't come up with one immediately. `canMergeDeclarationInto`?

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [reply mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:363]: I suspect that we may prefer cloning specific symbols from `other` in the longer run, as the entire module may be quite large and we don't necessarily need all of them. I'm fine with the current model if you an add the rationale above into the function documentation and something like `TODO: reconsider efficiency when we have enough users`.

File: mlir/lib/IR/SymbolTable.cpp
Comment (ftynse) [reply mlir/lib/IR/SymbolTable.cpp:236]: I doubt this is somehow intended, especially given that functions seem to share the implementation via templates. Can you investigate more?

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:172]: Why is this deleted?

**Review by ftynse — APPROVED** [review]

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ingomueller-net) [reply mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:172]: I wrote this in the commit message:

> Since the pass injects all definitions, providing the library again isn't needed. Since that injection isn't idempotent, it actually isn't even *possible* anymore, so this commits removes that argument.

I think no comment is needed in the code, here, right?

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ingomueller-net) [reply mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:301]: OK, done.

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [reply mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp]: Having a chain of commits, each with a small amount of code to avoid touching multiple subsystems. Here, it would be changing the SymbolTable in a separate commit. Some project follow this style, e.g., to have different code owners review them separately. LLVM/MLIR prefer having larger cross-cutting commits to see the overall motivation for changes.

No action is necessary (pull requests will be squashed into one commit anyway) so I resolved the discussion.

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ftynse) [reply mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp]: If this is short-lived, we can keep it. Basically, verifier already runs between passes and is intended for that, so I would prefer not having it run _inside_ the pass. We will end up duplicating the pass infrastructure...

I was discussing with @nicolasvasilache on a related subject, and it looks like having a separate pass for resolving symbols will be beneficial for generating reproducers. Not for this PR though!

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ingomueller-net) [reply mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:302]: Yeah, that could work. (As you probably also felt, it's not perfect: the name suggest that it *has* to be a declaration but it actually checks *if* it is a declaration.)

One argument in favor of keeping it as is is that this is a formulation that generalizes: In the SPIR-V module combiner, they "merge" global variables and constants, and those are "mergeable" if they define the same thing. This generalization could be formalized into a `MergeableSymbol` interface (which extends `Symbol`) with two methods `canMergeInto` and `mergeInto`. At least for these two examples (SPIR-V and TD), that concept seems to work...

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ingomueller-net) [reply mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:363]: OK, done. The `TODO` says that cloning individual ops probably makes most sense if we combine it with some pruning technique of the symbols that should be merged.

File: mlir/lib/IR/SymbolTable.cpp
Comment (ingomueller-net) [reply mlir/lib/IR/SymbolTable.cpp:236]: OK, will do.

**Review by ingomueller-net — COMMENTED** [review]
Thanks for the review! Now all comments should be addressed. I'll merge once CI has passed. Should some of the minor things still bother you/us, then I can still submit a fix-up.

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ingomueller-net) [reply mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp]: OK, I added a comment with a TODO.

File: mlir/lib/IR/SymbolTable.cpp
Comment (ingomueller-net) [reply mlir/lib/IR/SymbolTable.cpp:236]: For the record: fixed by #68320.

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (qcolombet) [reply mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:302]: Kind of late to the party, but I agree that the meaning is difficult to grasp currently.
More specifically I believe the comment doesn't describe what mergeable means.

For me the confusing part is that merging implies that the result is bigger. So before reading the code I was wondering if we were doing some sort of function concatenation :).
\<digression>LLVM has an optimization pass called merge globals that puts global variables in the same global structure to allow faster symbol resolution and that's the kind of thing this function name brings to mind for me.\</digression>

Maybe `canCoalesceSymbols`?

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (qcolombet) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:307]: I'll add a comment about "metadata" merging rules (i.e., around consumed and readonly)

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (qcolombet) [inline mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:362]: I think this comment would be easier to read if we break it down into several sentences (or bullet points?).

File: mlir/test/Dialect/Transform/test-interpreter-external-symbol-def-invalid.mlir
Comment (qcolombet) [inline mlir/test/Dialect/Transform/test-interpreter-external-symbol-def-invalid.mlir:1]: Do we need a run line here?
If `lit` doesn't skip this file, maybe we could move all the "include" files in a sub-dir and tell lit to ignore these files.

Edit: I see that we have patterns of files like that already (e.g., `mlir/test/Dialect/Transform/test-interpreter-external-symbol-def.mlir`). I don't particularly like this kind of "fake" run line, but maybe we are actually testing something here?
@ftynse what do you think we should do?

**Review by qcolombet — COMMENTED** [review]
I told @ingomueller-net I was going to look at this, so here we are :).

Sorry for the delay. LGTM too, couple of comments inlined.

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ingomueller-net) [reply mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:302]: I am open to using a different verb or name. From what I understand from the dictionary, [unite](https://www.merriam-webster.com/dictionary/unite), [combine](https://www.merriam-webster.com/dictionary/combine), [merge](https://www.merriam-webster.com/dictionary/merge), and [coalesce](https://www.merriam-webster.com/dictionary/coalesce) all roughly mean "take several things and make a single one of them."

I guess that that is undoubtably what we do with the outer ops (and the result is indeed larger usually). I'd argue that it is *close* to what we do with the symbols as well but maybe not *exactly* the same, at least not what we do with functions currently. Another angle to look at what we do with symbols is that we "deduplicate" them. Maybe that can yield a name?

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ingomueller-net) [reply mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:307]: You *will* or you *would*? I.e., who makes that commit?

File: mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp
Comment (ingomueller-net) [reply mlir/lib/Dialect/Transform/Transforms/TransformInterpreterPassBase.cpp:362]: Yep, sounds good. Let's have the person do it that adds the comment you suggested above.

File: mlir/test/Dialect/Transform/test-interpreter-external-symbol-def-invalid.mlir
Comment (ingomueller-net) [reply mlir/test/Dialect/Transform/test-interpreter-external-symbol-def-invalid.mlir:1]: I copied this from other files and speculated that this is done to (1) make lit not complain and (2) maybe make a test fail if the syntax of this specific file gets out of date rather than relying a test of another file?

**Review by ingomueller-net — COMMENTED** [review]
Thanks @qcolombet! The suggestions about the comments makes sense. Who does them?

**Pull request merged**
